Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs

```diff
@@ -160,5 +166,91 @@ public void RespectAnnotationPrefixes()
                 "my.universal.namespace.BussyMcBusface", 
                 new IonSerializer().Serialize(new Bus()));
         }
+
+        [TestMethod]
+        public void DeserializeAnnotatedIonToParentClassNameMatchingAnnotation()
+        {
+            IIonReader reader = IonReaderBuilder.Build(TestObjects.truckIonText);
+
+            Vehicle truck = defaultSerializer.Deserialize<Vehicle>(reader);
+
+            AssertIsTruck(truck);
+        }
+
+        [TestMethod]
+        public void DeserializeAnnotatedIonToClassNameMatchingAnnotation()
+        {
+            IIonReader reader = IonReaderBuilder.Build(TestObjects.truckIonText);
+
+            Truck truck = defaultSerializer.Deserialize<Truck>(reader);
+
+            AssertIsTruck(truck);
+        }
+
+        [TestMethod]
+        public void DeserializeAnnotatedIonToClassNameNotMatchingAnnotation()
+        {
+            string annotatedIonText = "NonMatching:: { }";
+
+            IIonReader reader = IonReaderBuilder.Build(annotatedIonText);
+
+            Truck truck = defaultSerializer.Deserialize<Truck>(reader);
+
+            AssertIsTruck(truck);
+        }
+
+        [TestMethod]
+        public void DeserializeAnnotatedIonToClassNameWithAnnotatedTypeAssemblies()
+        {
+            IonSerializationOptions options = new IonSerializationOptions
+            {
+                AnnotatedTypeAssemblies = new string[]
+                {
+                        typeof(Truck).Assembly.GetName().Name
+                }
+            };
+
+            IIonReader reader = IonReaderBuilder.Build(TestObjects.truckIonText);
+
+            IonSerializer ionSerializer = new IonSerializer(options);
+            Vehicle truck = ionSerializer.Deserialize<Vehicle>(reader);
+
+            AssertIsTruck(truck);
+
+            // Ensure default loaded assemblies are not searched in if annotated type assemblies are given
+            options = new IonSerializationOptions
+            {
+                AnnotatedTypeAssemblies = new string[]
+                {
+                    "mockAssemblyName"
+                }
+            };
+
+            reader = IonReaderBuilder.Build(TestObjects.truckIonText);
+            ionSerializer = new IonSerializer(options);
+            truck = ionSerializer.Deserialize<Vehicle>(reader);
+
+            Assert.AreNotEqual(TestObjects.nativeTruck.ToString(), truck.ToString());
+        }
+
+        [TestMethod]
+        public void DeserializeMultipleAnnotatedIonToClassNameWithAnnotated()
+        {
+            // Multiple annotations are ignored, it will only pick the first one.
+            IIonValue ionTruck = valueFactory.NewEmptyStruct();
+            ionTruck.AddTypeAnnotation("Truck");
+            ionTruck.AddTypeAnnotation("SecondAnnotation");
```

Comment:
Minor comment, seems like it would be worthwhile to have a counterexample here where the annotations are reversed, since this should be loaded from Truck which would be impossible to differentiate between ignoring the second annotation and simply going through the annotations in order.

Reply:
Added a counterexample in the test